pkg/disttask/framework/taskexecutor/task_executor.go

```diff
@@ -307,7 +308,9 @@ func (e *BaseTaskExecutor) Run() {
 		// reset it when we get a subtask
 		checkInterval, noSubtaskCheckCnt = SubtaskCheckInterval, 0
 
-		if e.stepExec != nil && e.stepExec.GetStep() != subtask.Step {
+		if e.stepExec != nil &&
+			(e.stepExec.GetStep() != subtask.Step ||
+				e.stepCtx.Err() != nil) { // Previous step ctx is done, cleanup and use a new one.
```

Comment:
check `e.stepCtx.Err()` separately, we don't need to recreate step executor when the task is cancelled, we can continue and let L289 breaks it

and you need to handle the case in L146 by cancel a `subtask context`, not a `step context`